simple-examples/Pure.hs

```diff
@@ -0,0 +1,231 @@
+
+{-
+   Here, the argument is decomposed by the tuple data constructor into two
+   pieces. Since the whole first argument is linear, the tuple constructor
+   in linear haskell is linear -- i.e., notice the type of the
+   constructor:
+
+   (,) :: a #-> b #-> (a,b)
+
+   Now, this does not mean that if we have some non-linear function with
+   an argument (a,b) that `a` and `b` have to be consumed exactly once.
+
+   A linear arrow in a data constructor merely signifies that the argument
+   that preceedes that arrow must be used linearly if the input to a
+   linear function is this data type with this constructor. Here, since
+   `(,) x y` was the first input to linearSwap, which is a linear function
+   (meaning the first input is linear), the components `x` and `y` must
+   be used linearly.  With a non-linear function, this is not the case.
+
+   Consider the next function as an example.
+-}
+
+nonLinearSubsume :: (a,a) -> (a,a)
+nonLinearSubsume (x,_) = (x,x)
+
+{-
+   Notice that the function above could not have a linear arrow. If it
+   did, it would not compile. Why is this?  Well, the first argument is
+   linear, and the constructor is linear in both components.
+
+   (,) :: a #-> b #-> (a,b)
+
+   Again, this means the `a` and `b` must be used linearly.
+
+   Yet, in the body of the function, `a` is used twice and `b` is used
+   zero times.
+-}
+
+
+linearIdentity2 :: a #-> a
+linearIdentity2 x = linearIdentity x
+
+{-
+   To use an input linearly (or a component of an input), we can pass it
+   to another linear function. Here, since linearIdentity is linear, we
+   can be sure the term (linearIdentity x) consumes x exactly once.
+   Hence, all of linearIdentity2 consumes the input x exactly once.
+
+   If we replaced it with the original `id`, this would fail to type check
+   because `id` has the non-linear type `a -> a`.  Thus, GHC isn't sure
+   that `id` uses its input exactly once.  If `id` doesn't use it's input
+   exactly once, then linearIdentity2 won't use it's input exactly once,
+   violating it's type signature.
+
+   Now, this does not mean that merely using a linear function makes the
+   use of a variable linear. For instance, both of the two functions below
+   use their input exactly twice.
+-}
+
+nonLinearPair :: a -> (a,a)
+nonLinearPair x = (linearIdentity x, linearIdentity x)
+
+nonLinearPair2 :: a -> (a,a)
+nonLinearPair2 x = (x, linearIdentity x)
+
+{-
+   Hopefully, this gives a basic sense of the arithmatic of how much an
+   input to a function is "used".
+-}
+
+
+regularIdentity :: a -> a
+regularIdentity x = linearIdentity x
+
+{-
+   Of course, the fact that a function is linear makes no difference in
+   non-linear functions. So, non-linear functions can call linear
+   functions willy-nilly and they will work as expected. To state the
+   obvious, linear functions are regular functions but not all functions
+   are linear functions.
+-}
+
+
+(#.) :: (b #-> c) -> (a #-> b) -> (a #-> c)
```

Comment:
You could even have all the arrows be linear  here.

Reply:
Oh yes, I purposely made it simpler.